### PR TITLE
[feature][PolyhedralGrid] Make PolyhedralGrid fully coherent with the DUNE grid interface.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR make the implementation of PolyhedralGrid fully compliant with DUNE grid interface and enables DUNE grid checks. Also, some minor compilation errors are fixed for CpGrid and current DUNE master.

### CODE REVIEW

**Comment by dr-robertk** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
Do you want any reviewing or checking of the PolyhedralGrid parts? The CpGrid-related parts look fine.

**Comment by dr-robertk** [comment]
If you want to spend time that would be good, but I understand it's a big PR and I just squashed things together. This PR basically added a lot of things I had implemented over the last years in various different branches. The fact that test_polyhedralgrid passed the DUNE grid check gives enough confidence and that the implementation is correct. What may be interesting is the Geometry implementation where I use the DUNE geometries to get the local-global mappings. That may be something for CpGrid also.

**Comment by dr-robertk** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
> The fact that test_polyhedralgrid passed the DUNE grid check gives enough confidence and that the implementation is correct.

I agree that gives some confidence.

> What may be interesting is the Geometry implementation where I use the DUNE geometries to get the local-global mappings. That may be something for CpGrid also.

Yes, that is interesting. I am curious: how did you deal with the generic case when a cell has no reasonable reference element? How do you distinguish between those cases? (I'd be happy to read the code, given a pointer.)

File: opm/grid/polyhedralgrid/geometry.hh [hunk 1]
```diff
@@ -41,6 +45,70 @@ namespace Dune
     typedef Dune::FieldVector< ctype, coorddimension > GlobalCoordinate;
     typedef Dune::FieldVector< ctype, mydimension >    LocalCoordinate;
 
+    typedef typename Grid::Traits::ExtraData  ExtraData;
+    typedef typename Grid::Traits::template Codim<codimension>::EntitySeed EntitySeed;
+
```
Comment (dr-robertk) [inline opm/grid/polyhedralgrid/geometry.hh]: @atgeirr : You can start reading from here.

Comment (atgeirr) [reply opm/grid/polyhedralgrid/geometry.hh]: Thanks, I looked at it and saw some things that were new to me. I did notice a minor possible bug, will make a note in the diff.

File: opm/grid/polyhedralgrid/geometry.hh [hunk 2]
```diff
@@ -41,6 +45,70 @@ namespace Dune
...
+        Storage( ExtraData data, EntitySeed seed )
+          : data_( data ), seed_( seed )
+        {}
+
+        Storage( ExtraData data )
+          : data_( data ), seed_()
+        {}
+
+        ExtraData data() const { return data_; }
```
Comment (atgeirr) [inline opm/grid/polyhedralgrid/geometry.hh]: Should this rather return a const reference? I imagine ExtraData is not something you want to copy for every call to corner() or center()?

Comment (dr-robertk) [reply opm/grid/polyhedralgrid/geometry.hh]: ExtraData is of type Grid* so this is no problem (see grid.hh). EntitySeed is essentially the element index or similar so copying this is also fine.

**Comment by atgeirr** [comment]
jenkins build this please

File: opm/grid/polyhedralgrid/geometry.hh (hunk 2 above)
Comment (atgeirr) [reply opm/grid/polyhedralgrid/geometry.hh]: Ok, if you want this merged now that is fine with me.

**Comment by dr-robertk** [comment]
It's ok from my side and would nice if it could be merged.

**Comment by akva2** [comment]
this broke build on ubuntu xenial (dune 2.4), ref https://ci.opm-project.org/job/opm-nightly-xenial/164/console

**Pull request merged**